### PR TITLE
Avoid unbound PS1 failure

### LINKED ISSUE #1029: activate may fails due to unbound variables when run with set -eu
It seems that an old bug is still here: `PS1: unbound variable` when called with bash strict mode.

Clearly virtualenv needs a test using bash strict mode and an almost empty set of environment variables, so we avoid future regressions on this.

**Please note that this bug does reproduce with ANY supported Unix shell, not only `bash`, including `ksh` and `zsh`.**

Here is a simple way to replicate the bug:
```
#!/bin/bash
# same applies to any other bourne compatible shells (is not bash specific)
set -euox pipefail
pip install -U virtualenv
virtualenv xxx
unset PS1
source xxx/bin/activate
```

The workaround, while ugly, is to do prepend `PS=${PS:-} ` to the activate line, which defines PS as empty string when is not already defined, or keeps its value when is defined.

The same kind of [bug applies to Python version of venv](https://bugs.python.org/issue25351) and there is an already [open PR to fix it there too](https://github.com/python/cpython/pull/3804).

Please do not postpone/delay implementing a fix just because the same kind of bug exists in other places. Note that the **default expansion variable syntax is POSIX compatible** and not something new or fancy, the fact that this was not known to the original author should not be an excuse for not using it.

### PR DESCRIPTION
**Pull request by ssbarnea**
fixes #1029

Change-Id: I764297341dee69a01bd4044d6567dee1026499d2

### CODE REVIEW

**Comment by pfmoore** [comment]
The embedded copy of the script needs refreshing. Just changing the raw script isn't sufficient - see `bin/rebuild-script.py`.

**Comment by ssbarnea** [comment]
@pfmoore Thanks for the hint. I am looking now and I will also try to improve the testing, maybe I can do it in a single PR that also improves the testing.

File: virtualenv_embedded/activate.sh [hunk 1]
```diff
@@ -1,5 +1,6 @@
 # This file must be used with "source bin/activate" *from bash*
 # you cannot run it directly
+set -x
```
Comment (jakub-bochenski) [inline virtualenv_embedded/activate.sh]: Not sure if people would appreciate suddenly getting debug output from activate

If we go with it we should restore the previous setting on exit at least

File: virtualenv_embedded/activate.sh [hunk 2]
```diff
@@ -54,11 +60,12 @@ if ! [ -z "${PYTHONHOME+_}" ] ; then
 fi
 
 if [ -z "${VIRTUAL_ENV_DISABLE_PROMPT-}" ] ; then
-    _OLD_VIRTUAL_PS1="$PS1"
+    _OLD_VIRTUAL_PS1="${PS1:-}"
```
Comment (jakub-bochenski) [inline virtualenv_embedded/activate.sh]: I think it's better to go with `${PS1-}` we just want to guard against unset variables. Why add the side-effect of setting PS1?

And look: this construct is already employed in line 62 above --this should remove any concerns about portability

File: virtualenv_embedded/activate.sh [hunk 3]
```diff
@@ -54,11 +60,12 @@ if ! [ -z "${PYTHONHOME+_}" ] ; then
...
+    # __VIRTUAL_PROMPT__ is not a variable, is replaced at compile time
     if [ "x__VIRTUAL_PROMPT__" != x ] ; then
-        PS1="__VIRTUAL_PROMPT__$PS1"
+        PS1="__VIRTUAL_PROMPT__${PS1:-}"
```
Comment (jakub-bochenski) [inline virtualenv_embedded/activate.sh]: ditto

File: virtualenv_embedded/activate.sh [hunk 4]
```diff
@@ -54,11 +60,12 @@ if ! [ -z "${PYTHONHOME+_}" ] ; then
...
     else
-        PS1="(`basename \"$VIRTUAL_ENV\"`) $PS1"
+        PS1="(`basename \"$VIRTUAL_ENV\"`) ${PS1:-}"
```
Comment (jakub-bochenski) [inline virtualenv_embedded/activate.sh]: ditto

File: virtualenv_embedded/activate.sh (hunk 2 above)
Comment (ssbarnea) [reply virtualenv_embedded/activate.sh]: @jakub-bochenski Out of curiosity, is there a difference between `${FOO-}` and `${FOO:-}`? I used the `:-` for years and that was the first time I encountered the simple minus.

Comment (jakub-bochenski) [reply virtualenv_embedded/activate.sh]: Haste makes waste -- I was wrong :)

The difference between `:-` and `-` is in treatment of null values, none of those does an assignment (you need either `=` or `:=`). Reference: http://pubs.opengroup.org/onlinepubs/9699919799/utilities/V3_chap02.html#tag_18_06_02

Still I think it might be better to use `-` because
 - it's already used
 - it only affects unset variables and leaves null as null, so it's a tighter fit

**Comment by pradyunsg** [comment]
@ssbarnea Travis and AppVeyor do not approve.

**Comment by jakub-bochenski** [comment]
@pradyunsg please note this requires #1087 to be merged first

**Comment by piotr-dobrogost** [comment]
@pradyunsg 
Could you please help move this forward?
Taking into consideration there appears to be no active maintainer I think mixing a fix with other changes (improving testing) here does not help getting this merged…

**Comment by ssbarnea** [comment]
@piotr-dobrogost I think you need to comment/review https://github.com/pypa/virtualenv/pull/1087 and try to persuade the team on irc and mailing list to put some love into virtualenv.

About one year ago I started some discussions that didn't prove to be fruitless. Who knows, maybe this year we would be able to do something.
